internal/requirements/requirements.go

```diff
@@ -0,0 +1,157 @@
+package requirements
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"strings"
+
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/encoding/unicode"
+)
+
+type requirementsTxt struct {
+	encoder *encoding.Encoder // encoder for reencoding requiremnets.txt
+	lines   []string          // decoded requirements
+	crlf    bool              // detected crlf line termination
+}
+
+func Read(r io.Reader) (*requirementsTxt, error) {
+	data, err := io.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+
+	encoding := unicode.UTF8
+	for _, bom := range boms {
+		if bom.Detect(data) {
+			encoding = bom.Encoding()
+			break
+		}
+	}
+
+	decoder := encoding.NewDecoder()
+	data, err = decoder.Bytes(data)
+	if err != nil {
+		return nil, err
+	}
+
+	lines, crlf, err := readLines(data)
+	if err != nil {
+		return nil, err
+	}
+
+	return &requirementsTxt{
+		encoder: encoding.NewEncoder(),
+		lines:   lines,
+		crlf:    crlf,
+	}, nil
+}
+
+func (r *requirementsTxt) Write(w io.Writer) error {
```

Comment:
The concatenation can be done in memory, e.g. this way:
``` go
func (r *requirementsTxt) Write(w io.Writer) error {
	ew := r.encoder.Writer(w)

	lineEnding := "\n"
	if r.crlf {
		lineEnding = "\r\n"
	}

	data := strings.Join(r.lines, lineEnding)
	_, err := ew.Write([]byte(data))

	return err
}

```

Reply:
Good point! That's much nicer.

Reply:
I updated the code with your suggestion :smile: